llama-index-integrations/llms/llama-index-llms-ai21/llama_index/llms/ai21/base.py

```
@@ -216,6 +216,7 @@ def _prepare_chat_with_tools(
         chat_history: Optional[List[ChatMessage]] = None,
         verbose: bool = False,
         allow_parallel_tool_calls: bool = False,
+        tool_required: bool = False,  # ai21 does not support configuring the tool_choice
```

Should we raise an error if its set to True?

---

@logan-markewich My thinking is that this will frequently get set more by internal library code rather than user code. Like, we'll want to migrate `StructuredLLM` to set `tool_required=True` (it's currently setting `tool_choice="function_name"` or something and hoping for the best from the LLM implementation). Seemed better to have it maybe give a tool response rather than blow up, like it currently is

---

otherwise the alternative is to somehow advertise whether the LLM supports tool_required or not, and checking that before providing it, which seems like a lot of gymnastics for mostly just a few underused LLMs

---

That's fair!